lib/_http_client.js

```diff
@@ -325,12 +312,24 @@ function ClientRequest(input, options, cb) {
     this._last = true;
     this.shouldKeepAlive = false;
     if (typeof options.createConnection === 'function') {
-      const newSocket = options.createConnection(options, oncreate);
-      if (newSocket && !called) {
-        called = true;
-        this.onSocket(newSocket);
-      } else {
-        return;
+      const oncreate = once((err, socket) => {
+        if (err) {
+          process.nextTick(() => this.emit('error', err));
+        } else {
+          this.onSocket(socket);
+          this._deferToConnect(null, null, () => this._flush());
+        }
+      });
+
+      try {
+        const newSocket = options.createConnection(options, oncreate);
+        if (newSocket) {
+          oncreate(null, newSocket);
```

Comment:
Without a `return` here, `this._deferToConnect(null, null, () => this._flush());` may (unnecessarily) end up being executed twice for this case, once in `oncreate()` and at the bottom of this function.